Add Next.js to supported Workers frameworks (#17124)
cloudflare/cloudflare-docs: src/content/docs/workers/frameworks/framework-guides/nextjs.mdx

---
pcx_content_type: how-to
title: Next.js
head: []
description: Create an Next.js application and deploy it to Cloudflare Workers with Workers Assets.
---

import {
	Badge,
	Description,
	InlineBadge,
	Render,
	PackageManagers,
} from "~/components";

#### New apps

To create a new Next.js app, pre-configured to run on Cloudflare using [`@opennextjs/cloudflare`](https://opennext.js.org/cloudflare), run:

```
npm create cloudflare@latest --no-auto-update --experimental --framework next
```

> #### Existing Next.js apps

Reviewer: Should call out the wrangler version here

```suggestion
#### Existing Next.js apps

::: note
Minimum required Wrangler version: 3.78.10. Check your version by running `wrangler version`. To update Wrangler, refer to [Install/Update Wrangler](https://developers.cloudflare.com/workers/wrangler/install-and-update/).
:::
```

##### 1. Install @opennextjs/cloudflare

First, install [@opennextjs/cloudflare](https://www.npmjs.com/package/@opennextjs/cloudflare):

```sh
npm install --save-dev @opennextjs/cloudflare
```

##### 2. Add a `wrangler.toml` file

Then, add a [`wrangler.toml`](/workers/wrangler/configuration/) file to the root directory of your Next.js app:

```toml
main = ".worker-next/index.mjs"
name = "my-app"
compatibility_date = "2024-09-23"
compatibility_flags = ["nodejs_compat"]
experimental_assets = { directory = ".worker-next/assets", binding = "ASSETS" }
```

<Callout>
As shown above, you must enable the [`nodejs_compat` compatibility flag](/workers/runtime-apis/nodejs/) *and* set your [compatibility date](/workers/configuration/compatibility-dates/) to `2024-09-23` or later, in order for your Next.js app to work with @opennextjs/cloudflare.
</Callout>

`wrangler.toml` is where you configure your Worker and define what resources it can access via [bindings](/workers/runtime-apis/bindings/).

##### 3. Update `package.json`

Add the following to the scripts field of your `package.json` file:

```json

> "build:worker": "cloudflare", TODO

Reviewer: TODO here

Author: ```suggestion
"build:worker": "cloudflare",
```

To enable caching, you must:

##### Create a KV namespace

```
npx wrangler@latest kv namespace create NEXT_CACHE_WORKERS_KV
```

##### Add the KV namespace to your Worker

```
[[kv_namespaces]]
binding = "NEXT_CACHE_WORKERS_KV"
id = "<YOUR_NAMESPACE_ID>"
```

#### Set the name of the binding to `NEXT_CACHE_WORKERS_KV`

As shown above, the name of the binding that you configure for the KV namespace must be set to `NEXT_CACHE_WORKERS_KV`.

##### 5. Develop locally

You can continue to run `next dev` when developing locally.

In step 3, we also added the `npm run preview:worker`, which allows you to quickly preview your app running locally in the Workers runtime, rather than in Node.js. This allows you to test changes in the same runtime as your app will run in when deployed to Cloudflare.

##### 6. Deploy to Cloudflare Workers

Either deploy via the command line:

```sh
npm run deploy
```

Or [connect a GitHub or GitLab repository](/workers/ci-cd/), and Cloudflare will automatically build and deploy each pull request you merge to your production branch.

---

## Static assets

> You can serve static assets your Astro application by placing them in the `./public/` directory. This can be useful for resource files such as images, stylesheets, fonts, and manifests.

Reviewer: Astro

```suggestion
You can serve static assets your Next.js application by placing them in the `./public/` directory. This can be useful for resource files such as images, stylesheets, fonts, and manifests.
```

---

import {
	Badge,
	Description,
	InlineBadge,
	Render,
	PackageManagers,
} from "~/components";

#### New apps

To create a new Next.js app, pre-configured to run on Cloudflare using [`@opennextjs/cloudflare`](https://opennext.js.org/cloudflare), run:

```
npm create cloudflare@latest --no-auto-update --experimental --framework next
```

#### Existing Next.js apps

::: note
Minimum required Wrangler version: 3.78.10. Check your version by running `wrangler version`. To update Wrangler, refer to [Install/Update Wrangler](https://developers.cloudflare.com/workers/wrangler/install-and-update/).
:::

##### 1. Install @opennextjs/cloudflare

First, install [@opennextjs/cloudflare](https://www.npmjs.com/package/@opennextjs/cloudflare):

```sh
npm install --save-dev @opennextjs/cloudflare
```

##### 2. Add a `wrangler.toml` file

Then, add a [`wrangler.toml`](/workers/wrangler/configuration/) file to the root directory of your Next.js app:

```toml
main = ".worker-next/index.mjs"
name = "my-app"
compatibility_date = "2024-09-23"
compatibility_flags = ["nodejs_compat"]
experimental_assets = { directory = ".worker-next/assets", binding = "ASSETS" }
```

> <Callout>

Author: ```suggestion
:::note
```

As shown above, you must enable the [`nodejs_compat` compatibility flag](/workers/runtime-apis/nodejs/) *and* set your [compatibility date](/workers/configuration/compatibility-dates/) to `2024-09-23` or later, in order for your Next.js app to work with @opennextjs/cloudflare.

> </Callout>

Author: ```suggestion
:::
```
